src/poly1305.rs

```diff
@@ -16,31 +16,115 @@
 // TODO: enforce maximum input length.
 
 use {c, chacha, constant_time, error};
+use core;
+use polyfill::slice::u32_from_le_u8;
 
-pub struct SigningContext {
-    state: State,
+struct AlignedBlockState<'a> {
+    buf: [u8; BLOCK_STATE_SIZE + 7],
+    inner: &'a mut [u8; BLOCK_STATE_SIZE],
+}
+impl<'a> AlignedBlockState<'a> {
+    pub fn new(inner: &mut [u8; BLOCK_STATE_SIZE]) -> AlignedBlockState {
+        let mut st = AlignedBlockState {
+            buf: [0u8; BLOCK_STATE_SIZE + 7],
+            inner: inner,
+        };
+        AlignedBlockState::align(&mut st.buf).copy_from_slice(&st.inner[..]);
+        st
+    }
+    fn align(xs: &mut [u8; BLOCK_STATE_SIZE + 7]) -> &mut [u8; BLOCK_STATE_SIZE] {
+        let aligned_start = (xs.as_ptr() as usize + 7) & !7;
+        let offset = aligned_start - (xs.as_ptr() as usize);
+        slice_as_array_ref_mut!(&mut xs[offset..offset+BLOCK_STATE_SIZE], BLOCK_STATE_SIZE).unwrap()
+    }
+    pub fn aligned_buf(&mut self) -> &mut [u8; BLOCK_STATE_SIZE] {
+        AlignedBlockState::align(&mut self.buf)
+    }
+}
+impl<'a> Drop for AlignedBlockState<'a> {
```

Comment:
I would rather just have the code explicitly copy the state from the temporarily aligned copy to the context's `opaque` when it is done. This `Drop` business is not explicit enough.